File: clients/go/list_offsets_test.go
package volant_test

import (
	"errors"
	"net"
	"sync"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type listOffsetsServer struct {
	mu             sync.Mutex
	errorCodes     []uint16
	entries        []codec.OffsetListing
	errorAsOpcode  bool
	meta           codec.MetadataResponse
	topic          string
	partitions     []uint32
	opcodes        []uint16
	listCount      int
	metadataCount  int
	acceptCount    int
	result         *listOffsetsServerResult
	err            error
	ln             net.Listener
}

type listOffsetsServerResult struct {
	topic      string
	partitions []uint32
	err        error
}

func startListOffsets(t *testing.T, errorCode uint16, entries []codec.OffsetListing) (*listOffsetsServer, string, func()) {
	t.Helper()
	return startListOffsetsCodes(t, []uint16{errorCode}, entries, false)
}

func startListOffsetsCodes(t *testing.T, errorCodes []uint16, entries []codec.OffsetListing, errorAsOpcode bool) (*listOffsetsServer, string, func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	codes := make([]uint16, len(errorCodes))
	copy(codes, errorCodes)
	s := &listOffsetsServer{
		errorCodes:    codes,
		entries:       entries,
		errorAsOpcode: errorAsOpcode,
		ln:            ln,
	}
	done := make(chan struct{})
	go func() {
		defer close(done)
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			s.mu.Lock()
			s.acceptCount++
			s.mu.Unlock()
			go s.serve(conn)
		}
	}()
	return s, ln.Addr().String(), func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func serveListOffsets(t *testing.T, errorCode uint16, entries []codec.OffsetListing) (addr string, got *listOffsetsServerResult, stop func()) {
	t.Helper()
	s, addr, innerStop := startListOffsets(t, errorCode, entries)
	got = &listOffsetsServerResult{}
	s.mu.Lock()
	s.result = got
	s.mu.Unlock()
	return addr, got, innerStop
}

func (s *listOffsetsServer) port() int {
	return s.ln.Addr().(*net.TCPAddr).Port
}

func (s *listOffsetsServer) snapshot() (lists, metas int, partitions []uint32, topic string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	parts := make([]uint32, len(s.partitions))
	copy(parts, s.partitions)
	return s.listCount, s.metadataCount, parts, s.topic
}

func (s *listOffsetsServer) serve(conn net.Conn) {
	defer conn.Close()
	_ = conn.SetDeadline(time.Now().Add(10 * time.Second))
	buf := []byte{}
	tmp := make([]byte, 4096)
	for {
		f, rest, err := frame.TryDecode(buf)
		if err != nil {
			s.mu.Lock()
			s.err = err
			s.mu.Unlock()
			return
		}
		if f == nil {
			n, err := conn.Read(tmp)
			if n > 0 {
				buf = append(buf, tmp[:n]...)
			}
			if err != nil {
				return
			}
			continue
		}
		buf = append([]byte(nil), rest...)
		raw, err := s.handle(f)
		if err != nil {
			s.mu.Lock()
			s.err = err
			s.mu.Unlock()
			return
		}
		if _, err := conn.Write(raw); err != nil {
			return
		}
		_ = conn.SetDeadline(time.Now().Add(10 * time.Second))
	}
}

func (s *listOffsetsServer) handle(f *frame.Frame) ([]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.opcodes = append(s.opcodes, f.Opcode)
	var payload []byte
	var err error
	replyOp := f.Opcode
	switch f.Opcode {
	case codec.OpListOffsets:
		s.listCount++
		req, e := codec.DecodeListOffsetsRequest(f.Payload)
		if e != nil {
			return nil, e
		}
		s.topic = req.Topic
		s.partitions = req.Partitions
		if s.result != nil {
			s.result.topic = req.Topic
			s.result.partitions = append([]uint32(nil), req.Partitions...)
		}
		code := uint16(0)
		if len(s.errorCodes) > 0 {
			code = s.errorCodes[0]
			s.errorCodes = s.errorCodes[1:]
		}
		if s.errorAsOpcode && code != 0 {
			payload, err = codec.EncodeErrorResponse(codec.ErrorResponse{Code: code})
			replyOp = codec.OpError
			break
		}
		entries := s.entries
		if code != 0 {
			entries = nil
		}
		payload, err = codec.EncodeListOffsetsResponse(codec.ListOffsetsResponse{
			ErrorCode: code,
			Topic:     req.Topic,
			Entries:   entries,
		})
		replyOp = codec.OpListOffsetsResponse
	case codec.OpMetadata:
		s.metadataCount++
		payload, err = codec.EncodeMetadataResponse(s.meta)
	default:
		return nil, &frame.ProtocolError{Msg: "unexpected opcode"}
	}
	if err != nil {
		return nil, err
	}
	return frame.Encode(replyOp, f.CorrelationID, payload)
}

func TestListOffsetsAllEncodesEmptyPartitions(t *testing.T) {
	entries := []codec.OffsetListing{
		{Partition: 0, Earliest: 0, Latest: 10},
		{Partition: 1, Earliest: 2, Latest: 5},
	}
	addr, got, stop := serveListOffsets(t, 0, entries)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ListOffsetsAll("events")
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || len(got.partitions) != 0 {
		t.Fatalf("wire request topic=%q partitions=%v", got.topic, got.partitions)
	}
	if len(out) != 2 || out[0] != (volant.OffsetListing{Partition: 0, Earliest: 0, Latest: 10}) {
		t.Fatalf("parsed %+v", out)
	}
	if out[1] != (volant.OffsetListing{Partition: 1, Earliest: 2, Latest: 5}) {
		t.Fatalf("parsed entry1 %+v", out[1])
	}
}

func TestListOffsetsEmptyPartitionsEncodedAsCountZero(t *testing.T) {
	entries := []codec.OffsetListing{
		{Partition: 0, Earliest: 0, Latest: 10},
		{Partition: 1, Earliest: 2, Latest: 5},
	}
	addr, got, stop := serveListOffsets(t, 0, entries)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ListOffsets("events", nil)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || len(got.partitions) != 0 {
		t.Fatalf("wire request topic=%q partitions=%v", got.topic, got.partitions)
	}
	if len(out) != 2 || out[0] != (volant.OffsetListing{Partition: 0, Earliest: 0, Latest: 10}) {
		t.Fatalf("parsed %+v", out)
	}
	if out[1] != (volant.OffsetListing{Partition: 1, Earliest: 2, Latest: 5}) {
		t.Fatalf("parsed entry1 %+v", out[1])
	}
}

func TestListOffsetsExplicitPartitions(t *testing.T) {
	entries := []codec.OffsetListing{{Partition: 0, Earliest: 0, Latest: 10}}
	addr, got, stop := serveListOffsets(t, 0, entries)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ListOffsets("events", []uint32{0, 1})
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if len(got.partitions) != 2 || got.partitions[0] != 0 || got.partitions[1] != 1 {
		t.Fatalf("partitions %v", got.partitions)
	}
	if len(out) != 1 || out[0].Latest != 10 {
		t.Fatalf("parsed %+v", out)
	}
}

func TestListOffsetsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveListOffsets(t, 2, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.ListOffsets("missing", nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "list_offsets" {
		t.Fatalf("broker error %+v", be)
	}
}

func TestListOffsetsError13RedirectsToLeader(t *testing.T) {
	entries := []codec.OffsetListing{{Partition: 0, Earliest: 0, Latest: 10}}
	leader, _, stopL := startListOffsets(t, 0, entries)
	defer stopL()
	follower, faddr, stopF := startListOffsetsCodes(t, []uint16{13}, nil, true)
	defer stopF()
	follower.mu.Lock()
	follower.meta = leaderMeta("events", 0, 2, "127.0.0.1", leader.port())
	follower.mu.Unlock()

	c, err := volant.DialTimeout(faddr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ListOffsets("events", nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(out) != 1 || out[0] != (volant.OffsetListing{Partition: 0, Earliest: 0, Latest: 10}) {
		t.Fatalf("parsed %+v", out)
	}
	fl, fm, _, _ := follower.snapshot()
	ll, _, parts, topic := leader.snapshot()
	if fl != 1 || fm != 1 {
		t.Fatalf("follower list/metadata = %d/%d want 1/1", fl, fm)
	}
	if ll != 1 || topic != "events" || len(parts) != 0 {
		t.Fatalf("leader list=%d topic=%s parts=%v", ll, topic, parts)
	}
}

func TestListOffsetsTypedError13RedirectsToLeader(t *testing.T) {
	entries := []codec.OffsetListing{{Partition: 0, Earliest: 0, Latest: 10}}
	leader, _, stopL := startListOffsets(t, 0, entries)
	defer stopL()
	follower, faddr, stopF := startListOffsets(t, 13, nil)
	defer stopF()
	follower.mu.Lock()
	follower.meta = leaderMeta("events", 0, 2, "127.0.0.1", leader.port())
	follower.mu.Unlock()

	c, err := volant.DialTimeout(faddr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.ListOffsets("events", nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(out) != 1 || out[0].Latest != 10 {
		t.Fatalf("parsed %+v", out)
	}
	fl, fm, _, _ := follower.snapshot()
	ll, _, _, _ := leader.snapshot()
	if fl != 1 || fm != 1 {
		t.Fatalf("follower list/metadata = %d/%d want 1/1", fl, fm)
	}
	if ll != 1 {
		t.Fatalf("leader list=%d want 1", ll)
	}
}

func TestListOffsetsError13MaxRedirectsZeroRaises(t *testing.T) {
	srv, addr, stop := startListOffsets(t, 13, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	c.SetMaxRedirects(0)
	_, err = c.ListOffsets("events", nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 13 || be.Op != "list_offsets" {
		t.Fatalf("broker error %+v", be)
	}
	lists, metas, _, _ := srv.snapshot()
	if lists != 1 || metas != 0 {
		t.Fatalf("list/metadata = %d/%d want 1/0", lists, metas)
	}
}

func TestListOffsetsRetriesTimeoutThenOkNoMetadata(t *testing.T) {
	entries := []codec.OffsetListing{{Partition: 0, Earliest: 0, Latest: 10}}
	srv, addr, stop := startListOffsetsCodes(t, []uint16{7, 0}, entries, false)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	c.SetMaxRetries(2)
	c.SetRetryBackoff(0)
	out, err := c.ListOffsets("events", nil)
	if err != nil {
		t.Fatal(err)
	}
	if len(out) != 1 || out[0].Latest != 10 {
		t.Fatalf("parsed %+v", out)
	}
	lists, metas, _, _ := srv.snapshot()
	if lists != 2 || metas != 0 {
		t.Fatalf("list/metadata = %d/%d want 2/0", lists, metas)
	}
}

func TestListOffsetsNotFoundNotRetriedNoMetadata(t *testing.T) {
	srv, addr, stop := startListOffsets(t, 2, nil)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	c.SetMaxRetries(2)
	c.SetRetryBackoff(0)
	_, err = c.ListOffsets("missing", nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "list_offsets" {
		t.Fatalf("broker error %+v", be)
	}
	lists, metas, _, _ := srv.snapshot()
	if lists != 1 || metas != 0 {
		t.Fatalf("list/metadata = %d/%d want 1/0", lists, metas)
	}
}

File: clients/java/src/test/java/io/volant/ListOffsetsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** ListOffsets client tests against a scripted TCP broker (no live server). */
class ListOffsetsTest {
    private static final int NOT_LEADER = Client.NOT_LEADER_FOR_PARTITION;

    @Test
    void emptyPartitionsEncodedAsCountZero() throws Exception {
        List<OffsetListing> entries =
                Arrays.asList(new OffsetListing(0, 0, 10), new OffsetListing(1, 2, 5));
        try (ListOffsetsServer srv = ListOffsetsServer.ok(entries)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<OffsetListing> got = c.listOffsets("events");
                assertEquals(entries, got);
            }
            assertEquals("events", srv.topic.get());
            assertTrue(srv.partitions.isEmpty());
        }
    }

    @Test
    void explicitPartitionsRoundtrip() throws Exception {
        List<OffsetListing> entries = Collections.singletonList(new OffsetListing(0, 0, 10));
        try (ListOffsetsServer srv = ListOffsetsServer.ok(entries)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                List<OffsetListing> got = c.listOffsets("events", 0, 1);
                assertEquals(entries, got);
            }
            assertEquals(Arrays.asList(0, 1), srv.partitions);
        }
    }

    @Test
    void nonzeroErrorCodeRaises() throws Exception {
        try (ListOffsetsServer srv = ListOffsetsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex = assertThrows(BrokerException.class, () -> c.listOffsets("missing"));
                assertEquals(2, ex.code);
                assertEquals("list_offsets", ex.op);
            }
        }
    }

    @Test
    void error13RedirectsToLeader() throws Exception {
        List<OffsetListing> entries = Collections.singletonList(new OffsetListing(0, 0, 10));
        try (ListOffsetsServer leader = ListOffsetsServer.ok(entries);
                ListOffsetsServer follower = ListOffsetsServer.errorOpcode(13)) {
            follower.meta = leaderMeta("events", 0, 2, "127.0.0.1", leader.port);
            try (Client c = Client.connect("127.0.0.1", follower.port, 5_000)) {
                List<OffsetListing> got = c.listOffsets("events");
                assertEquals(entries, got);
                assertEquals("127.0.0.1:" + leader.port, c.addr());
            }
            assertEquals(1, follower.listCount.get());
            assertEquals(1, follower.metadataCount.get());
            assertEquals(1, leader.listCount.get());
            assertEquals("events", leader.topic.get());
            assertTrue(leader.partitions.isEmpty());
        }
    }

    @Test
    void typedError13RedirectsToLeader() throws Exception {
        List<OffsetListing> entries = Collections.singletonList(new OffsetListing(0, 0, 10));
        try (ListOffsetsServer leader = ListOffsetsServer.ok(entries);
                ListOffsetsServer follower = ListOffsetsServer.error(13)) {
            follower.meta = leaderMeta("events", 0, 2, "127.0.0.1", leader.port);
            try (Client c = Client.connect("127.0.0.1", follower.port, 5_000)) {
                List<OffsetListing> got = c.listOffsets("events");
                assertEquals(entries, got);
                assertEquals("127.0.0.1:" + leader.port, c.addr());
            }
            assertEquals(1, follower.listCount.get());
            assertEquals(1, follower.metadataCount.get());
            assertEquals(1, leader.listCount.get());
            assertEquals(List.of(Codec.OP_LIST_OFFSETS, Codec.OP_METADATA), follower.opcodes);
        }
    }

    @Test
    void error13MaxRedirectsZeroRaises() throws Exception {
        try (ListOffsetsServer srv = ListOffsetsServer.error(13)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.setMaxRedirects(0);
                BrokerException ex =
                        assertThrows(BrokerException.class, () -> c.listOffsets("events"));
                assertEquals(NOT_LEADER, ex.code);
                assertEquals("list_offsets", ex.op);
            }
            assertEquals(List.of(Codec.OP_LIST_OFFSETS), srv.opcodes);
            assertEquals(1, srv.listCount.get());
            assertEquals(0, srv.metadataCount.get());
        }
    }

    @Test
    void retriesTimeoutThenOkNoMetadata() throws Exception {
        List<OffsetListing> entries = Collections.singletonList(new OffsetListing(0, 0, 10));
        try (ListOffsetsServer srv = ListOffsetsServer.codes(entries, 7, 0)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.setMaxRetries(2);
                c.setRetryBackoffMs(0);
                List<OffsetListing> got = c.listOffsets("events");
                assertEquals(entries, got);
            }
            assertEquals(List.of(Codec.OP_LIST_OFFSETS, Codec.OP_LIST_OFFSETS), srv.opcodes);
            assertEquals(2, srv.listCount.get());
            assertEquals(0, srv.metadataCount.get());
        }
    }

    @Test
    void notFoundNotRetried() throws Exception {
        try (ListOffsetsServer srv = ListOffsetsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                c.setMaxRetries(2);
                c.setRetryBackoffMs(0);
                BrokerException ex =
                        assertThrows(BrokerException.class, () -> c.listOffsets("missing"));
                assertEquals(2, ex.code);
                assertEquals("list_offsets", ex.op);
            }
            assertEquals(List.of(Codec.OP_LIST_OFFSETS), srv.opcodes);
            assertEquals(1, srv.listCount.get());
            assertEquals(0, srv.metadataCount.get());
        }
    }

    private static Metadata leaderMeta(String topic, int partition, int leaderId, String host, int port) {
        List<Metadata.BrokerInfo> brokers = new ArrayList<>();
        brokers.add(new Metadata.BrokerInfo(1, "127.0.0.1", 1));
        brokers.add(new Metadata.BrokerInfo(leaderId, host, port));
        return new Metadata(
                brokers,
                Collections.singletonList(new Metadata.TopicInfo(
                        topic,
                        1,
                        0,
                        Collections.singletonList(new Metadata.PartitionInfo(
                                partition,
                                leaderId,
                                0,
                                List.of(1L, (long) leaderId),
                                Collections.singletonList((long) leaderId),
                                1)))));
    }

    private static final class ListOffsetsServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> topic = new AtomicReference<>();
        final List<Integer> partitions = new CopyOnWriteArrayList<>();
        final List<Integer> opcodes = new CopyOnWriteArrayList<>();
        final List<Integer> errorCodes = new CopyOnWriteArrayList<>();
        final AtomicInteger listCount = new AtomicInteger();
        final AtomicInteger metadataCount = new AtomicInteger();
        final AtomicInteger acceptCount = new AtomicInteger();
        volatile Metadata meta = new Metadata(Collections.emptyList(), Collections.emptyList());
        private final List<OffsetListing> entries;
        private final boolean errorAsOpcode;
        private final ServerSocket listen;
        private final Thread acceptThread;

        static ListOffsetsServer ok(List<OffsetListing> entries) throws IOException {
            return new ListOffsetsServer(entries, false, 0);
        }

        static ListOffsetsServer error(int code) throws IOException {
            return new ListOffsetsServer(Collections.emptyList(), false, code);
        }

        static ListOffsetsServer errorOpcode(int code) throws IOException {
            return new ListOffsetsServer(Collections.emptyList(), true, code);
        }

        static ListOffsetsServer codes(List<OffsetListing> entries, int... codes) throws IOException {
            return new ListOffsetsServer(entries, false, codes);
        }

        private ListOffsetsServer(List<OffsetListing> entries, boolean errorAsOpcode, int... codes)
                throws IOException {
            this.entries = entries;
            this.errorAsOpcode = errorAsOpcode;
            if (codes.length == 0) {
                this.errorCodes.add(0);
            } else {
                for (int code : codes) {
                    this.errorCodes.add(code);
                }
            }
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            acceptThread = new Thread(
                    () -> {
                        while (!listen.isClosed()) {
                            try {
                                Socket conn = listen.accept();
                                acceptCount.incrementAndGet();
                                Thread t = new Thread(() -> serve(conn), "volant-list-offsets-conn");
                                t.setDaemon(true);
                                t.start();
                            } catch (IOException e) {
                                return;
                            }
                        }
                    },
                    "volant-list-offsets-accept");
            acceptThread.setDaemon(true);
            acceptThread.start();
        }

        private void serve(Socket conn) {
            try {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    int[] replyOp = new int[1];
                    byte[] payload = handle(d.frame, replyOp);
                    out.write(Frame.encode(replyOp[0], d.frame.correlationId, payload));
                    out.flush();
                }
            } catch (Exception ignored) {
                // client closed or timeout
            } finally {
                try {
                    conn.close();
                } catch (IOException ignored) {
                    // best-effort
                }
            }
        }

        private byte[] handle(Frame frame, int[] replyOp) {
            opcodes.add(frame.opcode);
            replyOp[0] = frame.opcode;
            if (frame.opcode == Codec.OP_LIST_OFFSETS) {
                listCount.incrementAndGet();
                Codec.ListOffsetsRequest req = Codec.decodeListOffsetsRequest(frame.payload);
                topic.set(req.topic);
                partitions.clear();
                partitions.addAll(req.partitions);
                int code = 0;
                if (!errorCodes.isEmpty()) {
                    code = errorCodes.remove(0);
                }
                if (errorAsOpcode && code != 0) {
                    replyOp[0] = Codec.OP_ERROR;
                    return Codec.encodeErrorResponse(new Codec.ErrorResponse(code, ""));
                }
                replyOp[0] = Codec.OP_LIST_OFFSETS_RESPONSE;
                return Codec.encodeListOffsetsResponse(
                        new Codec.ListOffsetsResponse(
                                code, req.topic, code == 0 ? entries : Collections.emptyList()));
            }
            if (frame.opcode == Codec.OP_METADATA) {
                metadataCount.incrementAndGet();
                return Codec.encodeMetadataResponse(meta);
            }
            throw new ProtocolException("unexpected opcode " + frame.opcode);
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                acceptThread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
